admittance_controller/src/admittance_controller.cpp

```diff
@@ -0,0 +1,602 @@
+  joint_state_ = last_reference_;
+
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::InterfaceConfiguration AdmittanceController::command_interface_configuration()
+  const
+{
+  if (!admittance_)
+  {
+    return {controller_interface::interface_configuration_type::INDIVIDUAL, {}};
+  }
+
+  std::vector<std::string> command_interfaces_config_names;
+  for (const auto & interface : admittance_->parameters_.command_interfaces)
+  {
+    for (const auto & joint : command_joint_names_)
+    {
+      auto full_name = joint + "/" + interface;
+      command_interfaces_config_names.push_back(full_name);
+    }
+  }
+
+  return {
+    controller_interface::interface_configuration_type::INDIVIDUAL,
+    command_interfaces_config_names};
+}
+
+controller_interface::InterfaceConfiguration AdmittanceController::state_interface_configuration()
+  const
+{
+  if (!admittance_)
+  {
+    return {controller_interface::interface_configuration_type::INDIVIDUAL, {}};
+  }
+
+  std::vector<std::string> state_interfaces_config_names;
+  for (size_t i = 0; i < admittance_->parameters_.state_interfaces.size(); ++i)
+  {
+    const auto & interface = admittance_->parameters_.state_interfaces[i];
+    for (const auto & joint : admittance_->parameters_.joints)
+    {
+      auto full_name = joint + "/" + interface;
+      state_interfaces_config_names.push_back(full_name);
+    }
+  }
+
+  auto ft_interfaces = force_torque_sensor_->get_state_interface_names();
+  state_interfaces_config_names.insert(
+    state_interfaces_config_names.end(), ft_interfaces.begin(), ft_interfaces.end());
+
+  return {
+    controller_interface::interface_configuration_type::INDIVIDUAL, state_interfaces_config_names};
+}
+
+std::vector<hardware_interface::CommandInterface>
+AdmittanceController::on_export_reference_interfaces()
+{
+  // create CommandInterface interfaces that other controllers will be able to chain with
+  if (!admittance_)
+  {
+    return {};
+  }
+
+  std::vector<hardware_interface::CommandInterface> chainable_command_interfaces;
+  auto num_chainable_interfaces = admittance_->parameters_.chainable_command_interfaces.size() *
+                                  admittance_->parameters_.joints.size();
+
+  // allocate dynamic memory
+  chainable_command_interfaces.reserve(num_chainable_interfaces);
+  reference_interfaces_.resize(num_chainable_interfaces, std::numeric_limits<double>::quiet_NaN());
+  position_reference_ = {};
+  velocity_reference_ = {};
+
+  // assign reference interfaces
+  auto index = 0ul;
+  for (const auto & interface : reference_interfaces_types_)
+  {
+    for (const auto & joint : admittance_->parameters_.joints)
+    {
+      if (hardware_interface::HW_IF_POSITION == interface)
+        position_reference_.emplace_back(reference_interfaces_[index]);
+      else if (hardware_interface::HW_IF_VELOCITY == interface)
+      {
+        velocity_reference_.emplace_back(reference_interfaces_[index]);
+      }
```

Comment:
Since `allowed_reference_interfaces_types_` is defined in the source code, we can ensure that the references are always in the right order. So we should not need to sort. However, it might be good idea to make a helper method or add an additional method to  `controller_interface::ChainableControllerInterface` so every controller that implements  the chainable interface does not need to have this logic.